Add basic ability to get latest version and install to installation library

dsplaisted opened the pull request:
Move enough implementation into the library to implement the following methods:

- `IDotnetInstaller.Install`
- `IDotnetReleaseInfoProvider.GetLatestVersion`

Usage:

```c#
        var releaseInfoProvider = InstallerFactory.CreateReleaseInfoProvider();
        var installer = InstallerFactory.CreateInstaller();

        var latestVersion = releaseInfoProvider.GetLatestVersion(InstallComponent.SDK, channel);

        installer.Install(
            new DotnetInstallRoot(targetInstallPath, InstallerUtilities.GetDefaultInstallArchitecture()),
            InstallComponent.SDK,
            latestVersion!);
```

Other interface methods are not hooked up and will throw `NotImplementedException`.

dsplaisted committed:
Remove DotnetVersion

dsplaisted committed:
Rename IBootstrapperController to IDotnetInstallManager

dsplaisted committed:
Delete one of the IDotnetInstaller interfaces

dsplaisted committed:
Move ArchiveDotnetInstaller and dependencies to library

dsplaisted committed:
Create public APIs in library for installation

dsplaisted committed:
Add library API to get latest release in a channel

nagilson committed:
Merge branch 'dnup' into library-implementation

nagilson commented on src/Installer/dnup/IDotnetInstaller.cs:
```diff
@@ -1,13 +0,0 @@
-
-namespace Microsoft.DotNet.Tools.Bootstrapper
-{
-    internal interface IDotnetInstaller
```
I'm tentative on removing this interface - I think the IDotnetInstaller interface with install, uninstall is a more helpful layer of abstraction for the library. However, I think this abstraction is helpful for dnup because it separates the steps of installation from one that doesn't need a mutex and the step that does. That could allow us to reuse all of the logic for an admin based installer in the future. (prepare would be downloading the exe/verifying it and commit would be running the installer.)

With that said, I wouldn't consider this blocking, it's very much a future concern and it's not super difficult to address.

nagilson replied:
Thanks, that makes sense to me.

nagilson reviewed (commented):
Looking good so far! I definitely think logging/telemetry is a top priority for us after this, as well as adding even more tests. I'll leave this feedback now. Working on adding a little more verification to the e2e tests, then ill help merge your PR into that pipeline change so we have a working PR CI flow.

nagilson committed:
Merge remote-tracking branch 'upstream/dnup' into library-implementation

nagilson committed:
fix  strong name issue

nagilson committed:
Allow public use of GetInstallArchitecture

we can chat if we want this to be internal - though it might be helpful for consumers to be able to marshal the architecture from the runtime into our architecture type that we use in our install objects, considering it's part of dotnet install root, which is also public

MichaelSimons commented on test/dnup.Tests/LibraryTests.cs:
```diff
@@ -0,0 +1,43 @@
+
+        Log.WriteLine($"Installing to path: {testEnv.InstallPath}");
+
+        installer.Install(
```
Do you see value in validating anything post install to validate the install method did what was intended?  e.g. did the sdk get installed to the specified directory.

nagilson replied:
Great suggestion ;) Yes, and that is what I'm doing here https://github.com/dotnet/sdk/pull/51405

nagilson commented on src/Installer/Microsoft.Dotnet.Installation/Internal/ArchiveDotnetExtractor.cs:
```diff
@@ -10,17 +10,17 @@
+namespace Microsoft.Dotnet.Installation.Internal;
 
-internal class ArchiveDotnetInstaller : IDotnetInstaller, IDisposable
+internal class ArchiveDotnetExtractor : IDisposable
```
I like this name! Agree it's an improvement.
